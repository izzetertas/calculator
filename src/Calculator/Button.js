import React from 'react'
import classnames from 'classnames'
const Button = ({ label, onClick, className }) => (
  <input
    type='button'
    onClick={onClick}
    value={label}
    className={classnames('button', className)}
  />
)

export default Button;
