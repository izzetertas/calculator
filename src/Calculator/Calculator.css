.container {
  margin: 50px auto;
  width: 320px;
  padding: 5px;
  border: 1px solid #1795d4;
}

.button {
  width: 24%;
  height: 50px;
  margin: 1px;
  border: 1px solid #1795d4;
}

.result {
  display: flex;
  width: 100%;
  text-align: right;
  flex-direction: column;
}

.output-container {
  display: flex;
  flex-direction: row;
}

.output {
  width: 96%;
  height: 44px;
  margin-right: 3px;
  font-size: 24px;
  text-align: right; 
}

.ac-button-wrapper {
  width: 33%;
  margin-right: 1px;
}

.ac-button-wrapper .button {
  width: 100%;
  background-color: #1795d4;
  color: white;
}

.label {
  font-size: 12px;
  text-align: right; 
  width: 98%;
  border: 1px solid #dacfcf;
  margin-bottom: 5px;
  padding: 3px 0;
  color: #d63232;
}
