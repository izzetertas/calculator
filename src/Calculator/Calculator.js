import React, { Component } from 'react'
import Button from './Button'
import CalculatorService from '../CalculatorService'
import './Calculator.css'

class Calculator extends Component {
  initialState = {
    history: '',
    output: '0',
    currentNumber: '0',
    firstNumber: null,
    operand: '',
    isClickedEqual: false,
    isPressedEqual: false,
    memoryRecordedNumber: 0
  }

  state = this.initialState
  
  handleOnClear = () => {
    this.setState(Object.assign({}, this.initialState, { memoryRecordedNumber: this.state.memoryRecordedNumber }))
  }

  get getinputs() {
    const { firstNumber, currentNumber } = this.state
    const arr = []
    if(firstNumber !== null) arr.push(firstNumber)
    if(currentNumber !== '') arr.push(parseFloat(currentNumber))
    return arr
  }

  handleOnMemoryButton = (e) => {
    const input = e.target.value;
    const { currentNumber, memoryRecordedNumber } = this.state
    if(input === 'MR') {
      this.setState({
        currentNumber: memoryRecordedNumber,
        output: memoryRecordedNumber
      })
    } else if(input === 'M+') {
      this.setState({
        memoryRecordedNumber : parseFloat(currentNumber) + memoryRecordedNumber
      })
    }

  }

  handleOnClickNumber = (e) => {
    const input = e.target.value;
    const { output, operand, currentNumber, isClickedEqual } = this.state
    if(output === 0 && input === '0') return

    if(isClickedEqual && !currentNumber) {
      return this.setState({
        currentNumber: input,
        firstNumber: null,
        operand: '',
        isClickedEqual: false,
        output: input,
      })
    }

    if(operand && !currentNumber) {
      return this.setState((previousState) => ({
        currentNumber: input,
        output: input,
      }))
    }
    this.setState((previousState) => ({
      currentNumber: previousState.currentNumber !== '0' ? previousState.currentNumber + input : input,
      output: previousState.currentNumber !== '0' ? parseFloat(previousState.currentNumber + input) : parseFloat(input),
    }))
  }

  handleOnClickEqual = (e) => {
    if(e.target.value !== '=') return
    
    const { firstNumber, currentNumber, operand, output } = this.state
    if(!operand || !firstNumber) return
    
    if(firstNumber !== null && !currentNumber) {
      const inputs = [firstNumber, firstNumber]
      const result = this.calculateResult(operand, inputs)
      return this.setState({
        isClickedEqual: true,
        currentNumber: '',
        output: result,
        history: this.getResultMessage(operand, inputs, output),
        firstNumber: parseFloat(result)
      })
    }
    const inputs = this.getinputs
    const result = this.calculateResult(operand, inputs)
    this.setState({
      isClickedEqual: true,
      currentNumber: '',
      output: result,
      history: this.getResultMessage(operand, inputs, output),
      firstNumber: parseFloat(result)
    })
  }

  handleOnClickOperand = (e) => {
    const input = e.target.value;
    const { output, operand, currentNumber, firstNumber, isClickedEqual } = this.state
      
    const inputs = this.getinputs
    if(!inputs) return

    if(!operand || isClickedEqual){  
      const newOperandParameterLength = CalculatorService.getOperandParameterLength(input)
      if(newOperandParameterLength > inputs.length)
        return this.setState({ 
          firstNumber: currentNumber ? parseFloat(currentNumber) : parseFloat(firstNumber),
          operand: input,
          currentNumber: '',
          isClickedEqual: false
        })

      if(newOperandParameterLength === inputs.length){
        const result = this.calculateResult(input, inputs)
        return this.setState({
          operand: input,
          firstNumber: parseFloat(result),
          currentNumber: '',
          output: result,
          history: this.getResultMessage(operand, inputs, output)
        })
      }
    }
    
    const operandParameterLength = CalculatorService.getOperandParameterLength(operand)
    if(operandParameterLength === inputs.length) {
      const result = this.calculateResult(operand, inputs)
      return this.setState({
        operand: input,
        firstNumber: parseFloat(result),
        currentNumber: '',
        output: result,
        history: this.getResultMessage(operand, inputs, output)
      })    
    }
    else if(operandParameterLength > inputs.length) {
      const newOperandParameterLength = CalculatorService.getOperandParameterLength(input)
      if(newOperandParameterLength > inputs.length) {
        return this.setState({ operand: input, currentNumber: '' })        
      }
      if(newOperandParameterLength === inputs.length) {
        const result = this.calculateResult(input, inputs)
        return this.setState({
          operand: input,
          firstNumber: parseFloat(result),
          currentNumber: '',
          output: result,
          history: this.getResultMessage(operand, inputs, output)
        })      
      }
    }
  }

  calculateResult = (operand, numbers) => {
    const calculator = new CalculatorService()
    return calculator.calculate(operand, ...numbers).toFixed(2)
  }
  

  getResultMessage = (operand, numbers) => `(${numbers[0]} ${operand} ${numbers[1] || ''})`

  render(){
    const { history, output } = this.state;

    return (
      <div className='container'>
        <div className='result'>
          {
            history && 
            <div className='label'>
              {history}
            </div>
          }
            
          <div className='output-container'>
            <div className='ac-button-wrapper'>
              <Button label={'AC'} onClick={this.handleOnClear} />
            </div>
            <div>
              <input type='text' value={output} className='output' />
            </div>
          </div>
        </div>
        <div>
          <Button label={'7'} onClick={this.handleOnClickNumber} />
          <Button label={'8'} onClick={this.handleOnClickNumber} />
          <Button label={'9'} onClick={this.handleOnClickNumber} />
          <Button label={'/'} onClick={this.handleOnClickOperand} />
        </div>
        <div>
          <Button label={'4'} onClick={this.handleOnClickNumber} />
          <Button label={'5'} onClick={this.handleOnClickNumber} />
          <Button label={'6'} onClick={this.handleOnClickNumber} />
          <Button label={'*'} onClick={this.handleOnClickOperand} />
        </div>
        <div>
          <Button label={'1'} onClick={this.handleOnClickNumber} />
          <Button label={'2'} onClick={this.handleOnClickNumber} />
          <Button label={'3'} onClick={this.handleOnClickNumber} />
          <Button label={'-'} onClick={this.handleOnClickOperand} />
        </div>
        <div>
          <Button label={'0'} onClick={this.handleOnClickNumber} />
          <Button label={'.'} onClick={this.handleOnClickNumber} />
          <Button label={'='} onClick={this.handleOnClickEqual} />
          <Button label={'+'} onClick={this.handleOnClickOperand} />
        </div>
        <div>
          <Button label={'MR'} onClick={this.handleOnMemoryButton} />
          <Button label={'M+'} onClick={this.handleOnMemoryButton} />
        </div>
      </div>
    )
  }
}

export default Calculator;