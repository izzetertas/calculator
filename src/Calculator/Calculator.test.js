import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import Calculator from './Calculator'

let initialState;
let wrapper;

describe('<Calculator/>', () => {
  beforeEach(() => {
    initialState = {
      history: '',
      output: '0',
      currentNumber: '0',
      firstNumber: null,
      operand: ''
    }
    wrapper = shallow(<Calculator />)
  })

  it('should render correctly', () => {
      // const wrapper = shallow(<Calculator />)
      expect(toJson(wrapper)).toMatchSnapshot();
      expect(wrapper.find('Button').length).toBe(17)
      expect(wrapper.find('input').find({type: 'text'}).text()).toEqual('')
  });

  it('should set initial state correctly', () => {
    expect(wrapper.state()).toEqual(initialState)
  });

  describe('when user clicks AC button', () => {
    it('should reset to initial state', () => {
      wrapper.setState({ history: '...' })
      wrapper.find('Button').find({ label : 'AC'}).simulate('click')
      expect(wrapper.state()).toEqual(initialState)
    });
  });

  it('should assign output correctly', () => {
    let component = wrapper.find('input').find({type: 'text'}) 
    expect(component.prop('value')).toEqual('0')
    expect(component.prop('className')).toEqual('output')
    const expected = '12.45'
    wrapper.setState({ output: expected })

    component = wrapper.find('input').find({type: 'text'}) 
    expect(component.prop('value')).toEqual(expected) 
  });
  
  it('should assign history correctly', () => {
    let component = wrapper.find('div').find({className: 'label'}) 
    expect(component.length).toBe(0)

    const expected = ' 2 * 12.5 = 25'
    wrapper.setState({ history: expected })

    component = wrapper.find('div').find({className: 'label'}) 
    expect(component.text()).toEqual(expected) 
  });

  describe('when click a number', () => {
    it('should assign output correctly', () => {
      const component = wrapper.find('Button').find({ label : '1'})
      component.simulate('click', { target: { value: '1' } })
      expect(wrapper.state('output')).toEqual(1)

      component.simulate('click', { target: { value: '2' } })
      expect(wrapper.state('output')).toEqual(12)
    });
  });

  describe('when click an operand', () => {
    it('should assign operand correctly', () => {
      const component = wrapper.find('Button').find({ label : '+'})
      component.simulate('click', { target: { value: '+' } })
      expect(wrapper.state('operand')).toEqual('+')
    });
  });

  describe('when click to equal button', () => {
    it('should calculate result and output correctly', () => {
      wrapper.setState({
        firstNumber: 10,
        currentNumber: '2',
        operand: '/'
      })
      const component = wrapper.find('Button').find({ label : '='})
      component.simulate('click', { target: { value: '=' } })
      expect(wrapper.state('output')).toEqual('5.00')
      expect(wrapper.state('history')).toEqual('(10 / 2)')
    });

  });
  
});