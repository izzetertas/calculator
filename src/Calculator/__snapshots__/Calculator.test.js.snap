// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Calculator/> should render correctly 1`] = `
<div
  className="container"
>
  <div
    className="result"
  >
    <div
      className="output-container"
    >
      <div
        className="ac-button-wrapper"
      >
        <Button
          label="AC"
          onClick={[Function]}
        />
      </div>
      <div>
        <input
          className="output"
          type="text"
          value="0"
        />
      </div>
    </div>
  </div>
  <div>
    <Button
      label="7"
      onClick={[Function]}
    />
    <Button
      label="8"
      onClick={[Function]}
    />
    <Button
      label="9"
      onClick={[Function]}
    />
    <Button
      label="/"
      onClick={[Function]}
    />
  </div>
  <div>
    <Button
      label="4"
      onClick={[Function]}
    />
    <Button
      label="5"
      onClick={[Function]}
    />
    <Button
      label="6"
      onClick={[Function]}
    />
    <Button
      label="*"
      onClick={[Function]}
    />
  </div>
  <div>
    <Button
      label="1"
      onClick={[Function]}
    />
    <Button
      label="2"
      onClick={[Function]}
    />
    <Button
      label="3"
      onClick={[Function]}
    />
    <Button
      label="-"
      onClick={[Function]}
    />
  </div>
  <div>
    <Button
      label="0"
      onClick={[Function]}
    />
    <Button
      label="."
      onClick={[Function]}
    />
    <Button
      label="="
      onClick={[Function]}
    />
    <Button
      label="+"
      onClick={[Function]}
    />
  </div>
  <div>
    <Button
      label="MR"
      onClick={[Function]}
    />
    <Button
      label="M+"
      onClick={[Function]}
    />
  </div>
</div>
`;
